Change \ to / in element field.

BrainforgeUK opened the pull request:
Pull Request for Issue #35015 .

### Summary of Changes

Added str_replace() to change \ to /.
Looks like only the str_replace() at line 189 maybe essential but included one at line 175 for consistency and clarity.

### Testing Instructions

Use a MS Windows hosted test site (may not be a problem on Linux sites).
Install this library:
[lib_bftest1.zip](https://github.com/joomla/joomla-cms/files/6915941/lib_bftest1.zip)

### Actual result BEFORE applying this Pull Request

The library installed correctly AND appears in the Extensions:Discover listing.
Can then be installed again from there - get 2 records in the extensions table!

Need to uninstall both before testing the patch.

### Expected result AFTER applying this Pull Request

The library installed correctly and does not appear in the Extensions:Discover listing.

### Documentation Changes Required

docs.joomla.org/Manifest_files
Has been updated with J4.0 library related information since originally written ...

Current documentation on manifest files for libraries is rather limited (no longer true!).
https://docs.joomla.org/Manifest_files

More information on this can be found elsewhere, such as ...
https://www.joomlashack.com/blog/how-tos/development/how-to-package-joomla-libraries/ (see comments at end)
https://joomla.stackexchange.com/questions/13681/manifest-file-format-for-installing-a-library-package

This pull request and previous pull request #34910 appear to complete an 'almost working' undocumented feature, namely ...

_What is not immediately obvious is that you can include a slash (/) in the folder name, so you can put all your company libraries into JPATH_SITE/libraries/mycompany using this:_
```
<libraryname>mycompany/mydosomething</libraryname>
<libraryname>mycompany/mydosomethingelse</libraryname>
```

BrainforgeUK committed:
Change \ to / in element field.

Issue #35015

RickR2H commented:
I have tested this item :white_check_mark: successfully on 3765cb86309192f4c862b92f2e2fa5cd9e2304f6<br /><br />Tested on a WAMP server and could reproduce the issue. The patch fixed the issue.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35023">issues.joomla.org/tracker/joomla-cms/35023</a>.</sub>

brianteeman commented:
Please fix the code style issues
https://ci.joomla.org/joomla/joomla-cms/46322/1/6

BrainforgeUK committed:
Changed multi-line statement spacing

BrainforgeUK commented:
Code style changes made - issue resolved?

Quy commented:
I have tested this item :white_check_mark: successfully on 07439b8e78124dd03c6512342f90d84a5f3126b8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35023">issues.joomla.org/tracker/joomla-cms/35023</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35023">issues.joomla.org/tracker/joomla-cms/35023</a>.</sub>

BrainforgeUK committed:
Corrected 'could not copy files' message (#5)

Library %s: Could not copy files from the source.
Library element field was not being populated.

BrainforgeUK committed:
Revert "Corrected 'could not copy files' message (#5)" (#6)

This reverts commit d384ac0c9e4b08e768a1be12e34a2c24556313d3.

PhilETaylor commented on administrator/components/com_installer/src/Model/DiscoverModel.php:
```diff
@@ -169,7 +169,14 @@ public function discover()
 		{
-			$key = implode(':', array($install->type, $install->element, $install->folder, $install->client_id));
+			$key = implode(':',
+				array(
+					$install->type,
+					str_replace('\\', '/', $install->element),
+					$install->folder,
+					$install->client_id
+				)
```
```suggestion
				[
					$install->type,
					str_replace('\\', '/', $install->element),
					$install->folder,
					$install->client_id
				]
```

PhilETaylor commented on administrator/components/com_installer/src/Model/DiscoverModel.php:
```diff
@@ -178,7 +185,14 @@ public function discover()
 			// Check if we have a match on the element
-			$key = implode(':', array($result->type, $result->element, $result->folder, $result->client_id));
+			$key = implode(':',
+				array(
+					$result->type,
+					str_replace('\\', '/', $result->element),
+					$result->folder,
+					$result->client_id
+				)
```
```suggestion
				[
					$result->type,
					str_replace('\\', '/', $result->element),
					$result->folder,
					$result->client_id
				]
```

BrainforgeUK replied:
Changed array() to [ ].

BrainforgeUK committed:
Changed array() to [ ].

wilsonge commented:
Thanks!
